### PR TITLE
Crow/0.1

### LINKED ISSUE #3064: [request] crow/0.1
### Package Details
  * Package Name/Version: **crow/0.1**
  * Website: **https://github.com/ipkn/crow**
  * Source code: **https://github.com/ipkn/crow**

### Description Of The Library / Tool

Crow is very fast and easy to use C++ micro web framework (inspired by Python Flask) 

Related to #621

### PR DESCRIPTION
**Pull request by madduci**
**crow/0.1**

closes #3064

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/crow/all/test_package/test_package.cpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#include "jwt/jwt.hpp"
```
Comment (uilianries) [inline recipes/crow/all/test_package/test_package.cpp:1]: IS it correct? I see "crow.h" as header for include: 

https://github.com/ipkn/crow/blob/master/examples/helloworld.cpp

File: recipes/crow/all/test_package/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup(TARGETS)
+
+find_package(cpp-jwt REQUIRED)
```
Comment (gocarlos) [inline recipes/crow/all/test_package/CMakeLists.txt:7]: needs to be changed

File: recipes/crow/all/test_package/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+add_executable(${PROJECT_NAME} test_package.cpp)
+target_link_libraries(${PROJECT_NAME} cpp-jwt::cpp-jwt)
```
Comment (gocarlos) [inline recipes/crow/all/test_package/CMakeLists.txt:10]: same here

File: recipes/crow/all/test_package/test_package.cpp (hunk 1 above)
Comment (madduci) [reply recipes/crow/all/test_package/test_package.cpp]: good catch, I've just used another project as template and forgot to update the example. Just updated the test

File: recipes/crow/all/test_package/CMakeLists.txt (hunk 2 above)
Comment (madduci) [reply recipes/crow/all/test_package/CMakeLists.txt]: good catch, I've just used another project as template and forgot to update the example. Just updated the test

File: recipes/crow/all/test_package/CMakeLists.txt (hunk 3 above)
Comment (madduci) [reply recipes/crow/all/test_package/CMakeLists.txt]: same as above

**Comment by madduci** [comment]
This library seems to not work with modern Boost (>=1.70.0), due to removal/deprecation of some methods in boost::asio. I'm trying with the oldest Boost possible available in the conan center, which is 1.69.0

File: recipes/crow/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+class CrowConan(ConanFile):
+    name = "crow"
+    homepage = "https://github.com/ipkn/crow"
+    description = "Crow is C++ microframework for web. (inspired by Python Flask)"
+    topics = ("conan", "web", "microframework", "header-only")
+    url = "https://github.com/conan-io/conan-center-index"
+    settings = "os", "compiler", "arch", "build_type"
+    generators = "cmake", "cmake_find_package", "cmake_find_package_multi"
+    exports_sources = ["patches/*"]
+    license = "BSD3"
```
Comment (prince-chrismc) [inline recipes/crow/all/conanfile.py:15]: https://spdx.org/licenses/BSD-3-Clause.html

```suggestion
    license = "BSD-3-Clause"
```

File: recipes/crow/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        tools.patch(**self.conan_data["patches"][self.version])
+
+    def _configure_cmake(self):
+        cmake = CMake(self)
+        cmake.configure(source_folder=self._source_subfolder)
+        return cmake
+
+    def build(self):
+        cmake = self._configure_cmake()
+        cmake.build()
```
Comment (prince-chrismc) [inline recipes/crow/all/conanfile.py:39]: Is this required? it's header only

Comment (prince-chrismc) [reply recipes/crow/all/conanfile.py:39]: The `_configure_cmake` is only used here if the answer is yes

File: recipes/crow/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    def requirements(self):
+        self.requires("openssl/1.1.1h")
+        self.requires("boost/1.69.0")
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = "crow-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+        tools.patch(**self.conan_data["patches"][self.version])
```
Comment (prince-chrismc) [inline recipes/crow/all/conanfile.py:30]: Should be done in build/package

File: recipes/crow/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
+from conans import ConanFile, tools, CMake
+import os
+
+required_conan_version = ">=1.28.0"
```
Comment (prince-chrismc) [inline recipes/crow/all/conanfile.py:4]: There's no components defines, I believe this can be removed

**Review by prince-chrismc — COMMENTED** [review]
https://github.com/ipkn/crow/blob/49edf898a5b9a39a0d69072cc2434c4f23692908/CMakeLists.txt#L37 Should we remove the forced cpp version?

File: recipes/crow/all/conanfile.py (hunk 5 above)
Comment (madduci) [reply recipes/crow/all/conanfile.py:39]: Yes, it creates an "amalgamate" header only file during the CMake process, this it is required.

File: recipes/crow/all/conanfile.py (hunk 6 above)
Comment (madduci) [reply recipes/crow/all/conanfile.py:30]: The problem is, the original CMake fails without the patch to exclude shipped tests and examples. If I do it in the package function, I cannot build the amalgamation header anymore

**Comment by prince-chrismc** [comment]
> This library seems to not work with modern Boost (>=1.70.0), due to removal/deprecation of some methods in boost::asio. I'm trying with the oldest Boost possible available in the conan center, which is 1.69.0

Can you put a comment in the code?

Even raise a `ConanInvalidConfiguration` if the version is too high to let the consumer know? I can not say I've seen that before but it would be nice... Thoughts?

File: recipes/crow/all/conanfile.py (hunk 6 above)
Comment (uilianries) [reply recipes/crow/all/conanfile.py:30]: move to build() method. All content from source folder will be copied to there. Use package folder only for header-only projects which doesn't require any build step

File: recipes/crow/all/conanfile.py (hunk 5 above)
Comment (uilianries) [reply recipes/crow/all/conanfile.py:39]: so remove no_copy_source, because as it suggests, won't copy sources files to build folder

Comment (prince-chrismc) [reply recipes/crow/all/conanfile.py:39]: Please add a comment, so we dont forget :smiley:

**Comment by madduci** [comment]
Sure, I am happy to add the raising of error if the version is too high, but is there actually a way to check which version of boost is being consumed, within the conanfile.py? Sorry for the question, but I don't know all the Conan internals and possibilities as you do. 

About the forced C++ version, I can remove the flag, but beware the code seems to be unmaintained since 2017, I don't know if there's so.e template magic that might be broken with newer or older C++ standards

**Comment by prince-chrismc** [comment]
> Sure, I am happy to add the raising of error if the version is too high, but is there actually a way to check which version of boost is being consumed, within the conanfile.py? Sorry for the question, but I don't know all the Conan internals and possibilities as you do.

I did not know how either but I found this https://github.com/conan-io/conan-center-index/blob/2b1db8eb35bb76c098bce1a58ce9448adce658b6/recipes/packio/all/test_package/conanfile.py#L11 searching the repo! let me know how it goes

> About the forced C++ version, I can remove the flag, but beware the code seems to be unmaintained since 2017, I don't know if there's some template magic that might be broken with newer or older C++ standards

I think we should let the CCI team answer... I suspect other projects may override the `cppstd` settings and doing so here is fine.

PS. templates are not magic! It's a very uncommon lesser known science which only obscure YouTube video teach the subject :stuck_out_tongue_closed_eyes:

**Comment by uilianries** [comment]
@madduci Consider https://github.com/madduci/conan-center-index/pull/1

**Review by uilianries — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

File: recipes/crow/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    settings = "os", "compiler", "arch", "build_type"
+    exports_sources = ["patches/*"]
+    license = "BSD-3-Clause"
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def requirements(self):
+        self.requires("boost/1.69.0")
```
Comment (SSE4) [inline recipes/crow/all/conanfile.py:19]: why old version?

Comment (madduci) [reply recipes/crow/all/conanfile.py:19]: Unfortunately there's an open issue https://github.com/ipkn/crow/issues/374

Boost 1.70 introduces changes but the library hasn't been updated yet

File: recipes/crow/all/conandata.yml [hunk 9]
```diff
@@ -0,0 +1,8 @@
+sources:
+  "0.1":
+    sha256: 140ca4a4d75ce5996cb103155580cb13b0b27082d1efbc331000a34af55b4390
+    url: https://github.com/ipkn/crow/archive/v0.1.tar.gz
+patches:
+  "0.1":
+    patch_file: "patches/001-fix-cmake-errors.patch"
+    base_path: "source_subfolder"
```
Comment (SpaceIm) [inline recipes/crow/all/conandata.yml:8]:
```suggestion
    - patch_file: "patches/001-fix-cmake-errors.patch"
      base_path: "source_subfolder"
```

File: recipes/crow/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = "crow-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def build(self):
+        if tools.Version(self.deps_cpp_info["boost"].version) >= "1.70.0":
+            raise ConanInvalidConfiguration("Crow requires Boost <1.70.0")
+
+        tools.patch(**self.conan_data["patches"][self.version])
```
Comment (SpaceIm) [inline recipes/crow/all/conanfile.py:30]:
```suggestion
        for patch in self.conan_data.get("patches", {}).get(self.version, []):
            tools.patch(**patch)
```

File: recipes/crow/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+        cmake = CMake(self)
+        cmake.configure(source_folder=self._source_subfolder)
+        cmake.build()
+
+    def package(self):
+        self.copy(pattern="LICENSE*", dst="licenses", src=self._source_subfolder)
+        self.copy("*.h", dst="include/crow", src="amalgamate")
```
Comment (SpaceIm) [inline recipes/crow/all/conanfile.py:37]:
```suggestion
        self.copy("*.h", dst=os.path.join("include", "crow"), src="amalgamate")
```

**Review by SpaceIm — CHANGES_REQUESTED** [review]

File: recipes/crow/all/conanfile.py (hunk 10 above)
Comment (SpaceIm) [reply recipes/crow/all/conanfile.py:30]: more flexible

**Review by prince-chrismc — APPROVED** [review]

**Review by SpaceIm — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Pull request merged**